Hide menu cursor when non-mouse (e.g. keyboard) input is received

frenzibyte opened the pull request:
- Closes https://github.com/ppy/osu/issues/9523

PR'ing this mainly for UX feedback, but so far it feels pretty good to me.

---

https://user-images.githubusercontent.com/22781491/195103032-e640d024-8b9e-492c-8c66-198c494cf6ec.mp4

frenzibyte committed:
Rename various cursor symbols

 - `IProvideCursor.MenuCursor` -> `IProvideCursor.Cursor`
 - `MenuCursor` -> `MenuCursorContainer

frenzibyte committed:
Expose bindable for identifying last input source

frenzibyte committed:
Enable NRT on `MenuCursorContainer`

frenzibyte committed:
Hide menu cursor when non-mouse input is applied

peppy commented on osu.Game/Graphics/Cursor/MenuCursorContainer.cs:
```diff
@@ -148,6 +168,8 @@ protected override void PopOut()
             activeCursor.ScaleTo(0.6f, 250, Easing.In);
         }
 
+        private void updateInternalVisibilityState(bool show) => fadeContainer.FadeTo(show ? 1 : 0, 120, Easing.OutQuint);
```
Rather than applying a second different fade, let's use the same one as in `PopIn` / `PopOut` (and move the logic to a shared point of calling).

frenzibyte replied:
I have played around with it, and while I've came up with a shared code that works well across different contexts, I can't say for sure if it's good or hacky:

<img width="861" alt="CleanShot 2022-10-12 at 17 47 29@2x" src="https://user-images.githubusercontent.com/22781491/195375016-200007fd-0161-44ee-bcce-d7a539a11207.png">

I've pushed it for now, but feel free to change it if you're not fond of it.

peppy replied:
I definitely was not fond of it. I've pushed a maybe better version, see what you think.

peppy reviewed (changes requested):
Requires DI attention for tests

peppy commented:
I think this is a good direction, but in addition I'd suggest:

- Hide the cursor after inactivity (similar to stable). Could potentially be done as a separate PR
- Stop hiding the cursor in rulesets that don't provide a cursor. With this change, I don't think that is required anymore. I personally disliked that logic, because it is possible to lose your cursor when in one of those rulesets:

https://user-images.githubusercontent.com/191335/195256088-5a82a429-74ba-4a2c-8bcb-36d5f93ff0a0.mp4

frenzibyte committed:
Refactor internal visibility logic to route back to `PopIn`/`PopOut`

frenzibyte committed:
Hide menu cursor when user is idle

frenzibyte committed:
Remove default invisible cursor container from rulesets

frenzibyte committed:
Make `OsuUserInputManager` dependency nullable for tournament client

frenzibyte committed:
Update xmlodc and add nullability attributes

frenzibyte committed:
Update existing usages of `Cursor` to handle null value

peppy committed:
Refactor visibility states to read better

peppy commented:
This feels like the kind of thing that would benefit from some basic test coverage, since it's hard to test every scenario manually. Want to add something simple?

frenzibyte committed:
Merge branch 'master' into hide-mouse-on-keyboard-input

frenzibyte committed:
Move mouse input detection inside `MenuCursorContainer` to allow testing

frenzibyte committed:
Add test cases for hiding cursor on keyboard input

frenzibyte committed:
Add test case for hiding cursor on idle

frenzibyte committed:
Hide manual input manager cursor in `OsuGameTestScene`s

Allows checking that the cursor is actually hidden in
`TestSceneOsuGame.TestCursorHidesWhenIdle`.

frenzibyte committed:
Fix broken test case
